Widgets more field types

joncppl opened the pull request:
I wanted DateTime and Int fields for the upcoming events widget, so I added them.

joncppl committed:
implement datetimefield, intfield, uintfield

joncppl committed:
missed in last commit

blx commented on dispatch/static/manager/src/js/components/fields/IntField.js:
```diff
@@ -0,0 +1,23 @@
+
+export default class IntField extends React.Component {
+  onChange(val) {
+    val = val.replace(/[^\d]/, '')
```
as opposed to `/[^\d]/g`?

without the global flag you can't guarantee that all non-numbers will be removed, only the first character. is that intentional?

```js
'fx123abc'.replace(/[^\d]/, '')
// 'x123abc'

'fx123abc'.replace(/[^\d]+/, '')
// '123abc'

'fx123abc'.replace(/[^\d]/g, '')
// '123'

blx commented on dispatch/tests/test_widget_fields.py:
```diff
@@ -1,7 +1,10 @@
 from dispatch.apps.events.models import Event
 from dispatch.theme import register
-from dispatch.theme.fields import CharField, TextField, ArticleField, ImageField, WidgetField, EventField, Field
+from dispatch.theme.fields import CharField, TextField, ArticleField, ImageField, \
```
there is also the weakly-arguably slightly nicer parenthesyntax
```py
from dispatch.theme.fields import (
    CharField, TextField, ...
    WidgetField, EventField, ...
)

blx commented on dispatch/tests/test_widget_fields.py:
```diff
@@ -691,3 +694,77 @@ def test_event_to_json_no_data(self):
+            testfield.validate(date_sz)
+            self.fail('Field date is invalid, exception should have been thrown')
+        except InvalidField:
+            pass
```
there is also built-in support in django.TestCase via unittest for expecting an exception:
```py
with self.assertRaises(InvalidField):
    testfield.validate(date_sz)
```

psiemens replied:
Nice, didn't know about this

blx commented on dispatch/theme/fields.py:
```diff
@@ -101,6 +103,44 @@ def validate(self, data):
+        val = super(UIntField, self).validate(data)
+
+        if val < 0:
+            raise InvalidField('%s must be greater than or equal to  0' % self.label)
```
extra space before 0?

psiemens commented on dispatch/static/manager/src/js/components/fields/IntField.js:
```diff
@@ -0,0 +1,23 @@
+
+import { TextInput } from '../inputs'
+
+export default class IntField extends React.Component {
```
I'd prefer to be little more verbose and call this `IntegerField`

psiemens commented on dispatch/static/manager/src/js/components/fields/IntField.js:
```diff
@@ -0,0 +1,23 @@
+        placeholder={this.props.label}
+        value={this.props.data || ''}
+        fill={true}
+        type="text"
```
You should be able to use `type="number"`, which I think removes the need for the regex validation above.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number

joncppl replied:
type="number" doesn't neatly support integers. 

ie. it allows decimal places, and doesn't call onChange when they are inserted

psiemens commented on dispatch/theme/fields.py:
```diff
@@ -101,6 +103,44 @@ def validate(self, data):
+            return data
+        return int(data, base=10)
+
+class UIntField(IntField):
```
Instead of using a separate field, could the `IntegerField` accept optional `min_value` and `max_value` keyword args? In my opinion, that would be a cleaner approach, and it's also the way Django configures its built-in db `IntegerField`: https://docs.djangoproject.com/en/1.11/ref/forms/fields/#integerfield

psiemens replied:
You should be able to accomplish that be overriding the `Field` constructor

joncppl committed:
use single, configurable IntegerField class

joncppl committed:
propagate and display field errors

joncppl committed:
chain update logic

joncppl committed:
add a sanity check to IntegerField bounds

joncppl committed:
update tests to match spec changes

joncppl commented:
I've addressed most of the concerns.
I agree that a single configurable IntegerField is a better option, so I made that change.

Also noticed that field errors weren't being displayed. That was an easy fix.

joncppl commented on dispatch/static/manager/src/js/components/fields/IntegerField.js:
```diff
@@ -0,0 +1,55 @@
+
+    val = val.replace(/[^\d-]+/g, '')
+
+    // allow a '-' at start, but nowhere else
```
As I am not a regex master, I could not find a way to implement this in a single line. Perhaps it is possible?

blx replied:
one option is something like this, which could be further reduced into one line if you really wanted.
```js
const prefix = val[0] === '-' ? '-' : ''
val = prefix + val.replace(/[^\d]/g, '')
```

joncppl commented on dispatch/static/manager/src/js/components/fields/IntegerField.js:
```diff
@@ -0,0 +1,55 @@
+
+  componentDidUpdate() {
+    //put the caret back in the correct place
+    this.refs.input.refs.input.setSelectionRange(
```
This just puts the text cursor back in the right place after using a regex to manipulate the contents. Otherwise it goes flying to the end when an invalid character is typed.

Wouldn't have to do all this if input's had a type="integer" or safari supported the pattern input attribute.

joncppl committed:
sanity check only necessary when both min and max values are set

psiemens commented on dispatch/static/manager/src/js/components/fields/IntegerField.js:
```diff
@@ -0,0 +1,55 @@
+        placeholder={this.props.label}
+        value={this.props.data || ''}
+        fill={true}
+        type="text"
```
Did you try using `type="number"` instead of text?

joncppl replied:
https://github.com/ubyssey/dispatch/pull/624#discussion_r128326556

I don't like how it still lets you type decimal points

psiemens replied:
Ah my bad, missed that

psiemens replied:
What about this? https://www.w3schools.com/tags/att_input_step.asp 

I've never used it before, but might be something worth looking into. I do like your current approach because it performs all the necessary validation before sending the data to the server.

joncppl replied:
I tried step. It makes little arrows that appear at the right hand side of the input field, which in addition to the arrow keys, (de)increment the value by step. However floating point values can still be entered manually.

Perhaps I am being overly pedantic; users will probably know they can't enter a fraction of an event and it will get caught by the server's field validation anyway.
